From: "Luis Claudio R. Goncalves" <lgoncalv@redhat.com>
Date: Wed, 20 Dec 2023 10:20:48 -0300
Subject: [PATCH 322/327] rt: mm/page_alloc: backport missing bits from
 __build_all_zonelists() fix
Origin: https://git.kernel.org/cgit/linux/kernel/git/rt/linux-stable-rt.git/commit?id=559de5bb32fbda54afb2a46a7cb50f1326e8e2de

A while ago upstream landed commit a2ebb51575828 ("mm/page_alloc: use
write_seqlock_irqsave() instead write_seqlock() + local_irq_save().")
to fix a problem that had already been worked on v5.10-rt via commit
7bdd3bd5143a4 ("Revert "mm/page_alloc: fix potential deadlock on
zonelist_update_seqseqlock""). Sebastian pointed out it was important
to backport the missing elements of a2ebb51575828 for code consistency.

Fixes: 7bdd3bd5143a4 ("Revert "mm/page_alloc: fix potential deadlock on zonelist_update_seqseqlock"")
Cc: stable-rt@vger.kernel.org
Suggested-by: Sebastian Andrzej Siewior <bigeasy@linutronix.de>
Signed-off-by: Luis Claudio R. Goncalves <lgoncalv@redhat.com>
---
 mm/page_alloc.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index b5d2e5b84afb..7ea4f151b019 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -6054,8 +6054,9 @@ static void __build_all_zonelists(void *data)
 	int nid;
 	int __maybe_unused cpu;
 	pg_data_t *self = data;
+	unsigned long flags;
 
-	write_seqlock(&zonelist_update_seq);
+	write_seqlock_irqsave(&zonelist_update_seq, flags);
 
 #ifdef CONFIG_NUMA
 	memset(node_load, 0, sizeof(node_load));
@@ -6088,7 +6089,7 @@ static void __build_all_zonelists(void *data)
 #endif
 	}
 
-	write_sequnlock(&zonelist_update_seq);
+	write_sequnlock_irqrestore(&zonelist_update_seq, flags);
 }
 
 static noinline void __init
